File: clients/java3/src/org/openamq/client/framing/AMQMethodBody.java
package org.openamq.client.framing;

import org.apache.mina.common.ByteBuffer;

public abstract class AMQMethodBody extends AMQBody
{
    public static final byte TYPE = 1;             
    
    public short flags;
    
    public short reserved;
    
    public long synchtag;
    
    /** unsigned short */
    protected abstract int getBodySize();
    
    protected abstract byte getClazz();
    
    protected abstract byte getMethod();
    
    protected abstract void writeMethodPayload(ByteBuffer buffer);
    
    protected byte getType()
    {
        return TYPE;
    }

    protected int getSize()
    { 
        return 1 + 1 + 1 + 1 + getBodySize();
    }
    
    protected void writePayload(ByteBuffer buffer)
    {
        buffer.put(getClazz());
        buffer.put(getMethod());
        EncodingUtils.writeUnsignedByte(buffer, flags);        
        EncodingUtils.writeUnsignedByte(buffer, reserved);                
        writeMethodPayload(buffer);
    }

    protected abstract void populateMethodBodyFromBuffer(ByteBuffer buffer) throws AMQFrameDecodingException;
    
    protected void populateFromBuffer(ByteBuffer buffer) throws AMQFrameDecodingException
    {
        flags = buffer.getUnsigned();
        reserved = buffer.getUnsigned();        
        populateMethodBodyFromBuffer(buffer);
    }
    
    public String toString()
    {
        StringBuffer buf = new StringBuffer(getClass().toString());
        buf.append("\n\tClass: ").append(getClazz());
        buf.append("\n\tMethod: ").append(getMethod());
        buf.append("\n\tFlags: ").append(flags);        
        buf.append("\n\tReserved: ").append(reserved);                
        return buf.toString();
    }
}

File: dotnet/xmscommon/framing/AMQMethodBody.cs
using System;
using System.Text;
using jpmorgan.mina.common;
using log4net;

namespace OpenAMQ.Framing
{
    public abstract class AMQMethodBody : IBody
    {
        public const int TYPE = 1;

        public uint _synchTag;

        public uint SynchTag
        {
            get
            {
                return _synchTag;
            }
            set
            {
                _synchTag = value;
            }
        }

        protected abstract uint BodySize
        {
            get;
        }

        protected abstract ushort Clazz
        {
            get;
        }

        protected abstract ushort Method
        {
            get;
        }

        protected abstract void WriteMethodPayload(ByteBuffer buffer);

        public byte BodyType
        {
            get
            {
                return TYPE;
            }
        }

        public uint Size
        {
            get
            {
                return (uint) (2 + 2 + BodySize);
            }
        }

        public void WritePayload(ByteBuffer buffer)
        {
            buffer.Put(Clazz);
            buffer.Put(Method);
            WriteMethodPayload(buffer);
        }

        /// <summary>
        /// Populates the method body by decoding the specified buffer
        /// </summary>
        /// <param name="buffer">The buffer to decode.</param>
        /// <exception cref="AMQFrameDecodingException">If the buffer cannot be decoded</exception>
        protected abstract void PopulateMethodBodyFromBuffer(ByteBuffer buffer);

        /// <summary>
        /// Populates this instance from a buffer of data.
        /// </summary>
        /// <param name="buffer">The buffer.</param>
        /// <param name="size">The size.</param>
        /// <exception cref="AMQFrameDecodingException">If the buffer contains data that cannot be decoded</exception>
        public void PopulateFromBuffer(ByteBuffer buffer, uint size)
        {
            PopulateMethodBodyFromBuffer(buffer);
        }

        public override string ToString()
        {
            StringBuilder buf = new StringBuilder(GetType().ToString());
            buf.Append("\n\tClass: ").Append(Clazz);
            buf.Append("\n\tMethod: ").Append(Method);
            return buf.ToString();
        }
    }
}
